### PR TITLE
RavenDB-20044 Optimizations inside `EntryIdEncodings`.

### PR DESCRIPTION
**Pull request by maciejaszyk**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-20044

### Additional description

I tried to implement more efficient methods using SIMD (AVX), but after running some benchmarks, some had worse performance than the initial one. Therefore, I decided to only push the methods that showed an increase in performance, and attach the source code of the remaining methods on YouTrack.

### Type of change

- Optimization

### How risky is the change?

- Not relevant

### Backward compatibility

- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing by Contributor

- Tests have been added that prove the fix is effective or that the feature works

### Testing by RavenDB QA team

- No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

### CODE REVIEW

File: src/Corax/Utils/EntryIdEncodings.cs [hunk 1]
```diff
@@ -51,44 +51,65 @@ public static long GetContainerId(long entryId)
...
-            entries[i] = ecd;
+            entryId = (entryId << EntryIdOffset) | (FrequencyQuantization(frequency) << ContainerTypeOffset);
         }
     }
 
     [MethodImpl(MethodImplOptions.AggressiveInlining | MethodImplOptions.AggressiveOptimization)]
     public static long DecodeAndDiscardFrequency(long entryId) => entryId >> EntryIdOffset;
 
+    [MethodImpl(MethodImplOptions.AggressiveOptimization)]
+    public static unsafe void DecodeAndDiscardFrequencySimd(Span<long> entries, int read)
```
Comment (ayende) [inline src/Corax/Utils/EntryIdEncodings.cs]: Let's add ARM SIMD as well here, we are likely going to start using this more in the future, and it is fairly simple SIMD code

Comment (redknightlois) [reply src/Corax/Utils/EntryIdEncodings.cs]: Agreed, we just need to have access to those types of boxes.

File: test/FastTests/Corax/Ranking/QuantizationTest.cs [hunk 2]
```diff
@@ -25,4 +27,23 @@ public void CanEncodeAndDecodeEveryNumberUnderShort()
...
+
+    public void SimdInstructionCorrectlyIgnoresFrequency(int size)
+    {
+        var random = new Random(2337);
+        var ids = Enumerable.Range(0, size).Select(i => (long)random.Next(31_111, 59_999)).ToArray();
+        var idsWithShifted = ids.Select(i => i << 10).ToArray();
+        
+        EntryIdEncodings.DecodeAndDiscardFrequencySimd(idsWithShifted.AsSpan(), size);
+        Assert.Equal(ids, idsWithShifted);
+    }
```
Comment (redknightlois) [inline test/FastTests/Corax/Ranking/QuantizationTest.cs:63]: Add also a fuzzying test for equality of SIMD and Classic implementation.

File: bench/Voron.Benchmark/Corax/CoraxEntryIdEncodings.cs [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        return _idsWithEncodings;
+    }
+    
+    [Benchmark]
+    public Span<long> ClassicDiscard()
+    {
+        EntryIdEncodings.DecodeAndDiscardFrequency(_idsWithEncodings, BufferSize);
+        return _idsWithEncodings;
+    }
+}
```
Comment (redknightlois) [inline bench/Voron.Benchmark/Corax/CoraxEntryIdEncodings.cs:57]: Post in the PR the results of this, so we can have the reference with the current hardware if we ever need to check this.

File: src/Corax/Utils/EntryIdEncodings.cs (hunk 1 above)
Comment (maciejaszyk) [reply src/Corax/Utils/EntryIdEncodings.cs]: Shouldn't we also write a vectorized `AND` for ARM?

Comment (redknightlois) [reply src/Corax/Utils/EntryIdEncodings.cs]: Eventually, Yes.

Comment (maciejaszyk) [reply src/Corax/Utils/EntryIdEncodings.cs]: Implemented. MacOS test from GH Action should test it.

File: test/FastTests/Corax/Ranking/QuantizationTest.cs (hunk 2 above)
Comment (maciejaszyk) [reply test/FastTests/Corax/Ranking/QuantizationTest.cs:63]: Added 👍

File: bench/Voron.Benchmark/Corax/CoraxEntryIdEncodings.cs (hunk 3 above)
Comment (arekpalinski) [reply bench/Voron.Benchmark/Corax/CoraxEntryIdEncodings.cs:57]: Benchmarks results were put in the issue

File: src/Corax/Utils/EntryIdEncodings.cs (hunk 1 above)
Comment (arekpalinski) [reply src/Corax/Utils/EntryIdEncodings.cs]: https://issues.hibernatingrhinos.com/issue/RavenDB-20103

**Pull request merged**